pysmt/smtlib/printers.py

```diff
@@ -23,14 +23,45 @@
 from pysmt.walkers import TreeWalker, DagWalker, handles
 from pysmt.utils import quote
 
+def write_annotations(f):
+    def resf(self, formula, *args, **kwargs):
+        if self.annotations is not None and formula in self.annotations:
+            self.write("(! ")
+        res = f(self, formula, *args, **kwargs)
+        # all of the items must be yielded to avoid printing out of order
+        if res is not None:
+            for item in res:
+                yield item
+        if self.annotations is not None and formula in self.annotations:
+            for key, values in self.annotations[formula].items():
+                self.write(f" :{key}")
+                for value in values:
+                    self.write(f" {value}")
+            self.write(')')
+    return resf
+
+def write_annotations_dag(f):
+    def resf(self, formula, *args, **kwargs):
+        if self.annotations is not None and formula in self.annotations:
+            res = ["(! "]
+            res.append(f(self, formula, *args, **kwargs))
+            for key, values in self.annotations[formula].items():
+                res.append(f" :{key}")
+                for value in values:
+                    res.append(f" {value}")
+            res.append(")")
+            return ''.join(res)
+        return f(self, formula, *args, **kwargs)
```

Comment:
A comment on nesting and indentation: by complementing the predicate of the `if` statement, eight lines can be moved to a smaller indentation level, and the maximal nesting depth of syntactic constructs reduced from 5 to 4. Together with other changes:

```python
def write_annotations_dag(f):
    def resf(self, formula, *arg, **kwargs):
        annots = self.annotations
        app = f(self, formula, *args, **kwargs)
        if annots is None or formula not in annots:
            return app
        items = list()
        for key, values in annots[formula].items():
            items.append(f' :{key}')
            items.extend(f' {v}' for v in values)
        kv = ''.join(items)
        return f'(! {app}{kv})'
```

Reply:
Thanks for the contribution @johnyf!
@agirardi-fbk can you have a look and update the code / comment here?

Reply:
Thank you for your suggestion @johnyf, I have updated the pull request

Reply:
Thank you all for reviewing these changes, and updating the pull request.